docs/proposals/provisioning-proposal.md

```diff
@@ -0,0 +1,112 @@
+# Cluster-API Provisioning Mechanism Consolidation Proposal
+
+## Abstract
+
+Cluster-API requires implementers to implement provisioning mechanisms (bootstrap scripts and logic for running those scripts) themselves. However, the mechanism, including bootstrap scripts, is same or very similar for all cloud providers. As this is a time consuming task it could be very useful to provide a reusable provisioning mechanism in the upstream Cluster-API.
+
+## Proposal
+
+The upstream Cluster-API doesn’t have any bootstrap scripts neither any wrapper around kubeadm for handling cluster operations (initialization and join). This means implementers have to manually write bootstrap scripts used by specific Cluster-API provider and manually handle cluster initialization and node joining. For most cloud providers both bootstrap scripts and steps for cluster initialization/joining are the same, except for provider specifics that could be provided via template variables.
+
+The bootstrap scripts usually do the following tasks:
+
+* Install dependencies required by Kubernetes
+    * Includes dependencies such as socat, ebtables, which are same for all providers.
+* Install container runtime
+    * Steps for installing container runtime such as Docker or containerd are same for all providers.
+* Install Kubernetes packages—kubeadm, kubelet, kubectl, kubernetes-cni
+    * The required packages and steps to install them are same for all providers.
+* Configure kubelet
+    * There are several differences based on provider, however differences could be templated.
+    * kubeadm configuration file can handle some configuration tasks for kubelet.
+* Initialize cluster or join node to a cluster
+    * Should be done using kubeadm configuration files, written by implementer or Cluster-API provider user.
+
+Implementing the provisioning scripts in the upstream Cluster-API ensures:
+
+* Cluster-API implementers can easily get started, as they don’t need to implement complex provisioning logic and scripts for many various setups and operating systems
+* The bootstrap scripts are located in one place, which makes it easier to maintain, distribute and test them
+
+### Example use case
+
+Han wants to write a new Cluster-API provider for his employers cloud solution. Han knows the only difference is the cloud providers’ API, hence Han only wants to implement those API calls. Because the Cluster-API provides templates for the actual provisioning, Han can easily import and re-use them and safely assume they will work, as they are well tested.
+
+## Goals
+
+* Cluster-API provides re-usable provisioning scripts for bootstrapping a cluster for the most popular operating systems (Ubuntu, CentOS and CoreOS)
```

Comment:
@roberthbailey: It feels like the scripts have two jobs, that we could maybe separate. There is the "install stuff" step, which I think could be argued is mostly the same across providers, and then there is the "configure and run stuff" part, which either needs heavy templating or shouldn't be shared. If we focused on the first part, how much would that go towards solving the issues you are hoping to address?

@roberthbailey: This is still my biggest outstanding question about this proposal. Re-reading the doc, it seems like we are going to end up with a lot of templating for the second part, which may not make it particularly easy for "Han" to pick up the scripts and just use them in a new environment.